File: answers/other/in sap/Additive Number.cpp
class Solution {
public:
    bool isAdditiveNumber(string num) {
        for (size_t i = 0; i < num.length(); ++ i) {
            if (isAdditiveNumber(0, i + 1, num, 0)) {
                return true;
            }
        }
        return false;
    }
    bool isAdditiveNumber(size_t begin, size_t current, const string &num, int level) {
        cout << "debug " << begin << " " << current << endl;
        if (current + 1 >= num.length() && level > 0) {
            return true;
        }
        for (size_t i = current; i < num.length(); ++ i) {
            string sum;
            size_t j = current - 1;
            size_t k = i;
            int carry = 0;
            while ((j >= begin && j < num.length()) || (k >= current && k < num.length()) || carry) {
                int digit = carry;
                if (j >= begin && j < num.length()) {
                    digit += num[j] - '0';
                }
                if (k >= current && k < num.length()) {
                    digit += num[k] - '0';
                }
                carry = digit / 10;
                sum = static_cast<char> (digit % 10 + '0') + sum;
                -- j;
                -- k;
            }
            cout << "check " << sum << " " << num.substr(i + 1, sum.length()) << endl;
            if (sum == num.substr(i + 1, sum.length()) && (i + 1 + sum.length() >= num.length() || isAdditiveNumber(current, i + 1, num, level + 1))) {
                return true;
            }
        }
        return false;
    }
};

File: answers/leetcode/Additive Number/Additive Number.java
class Solution {
    public boolean isAdditiveNumber(String num) {
        if (num == null || num.length() == 0) {
            return false;
        }
        for (int i = 1; i + 1 < num.length(); i++) {
            if (this.dfs(0, i, num)) {
                return true;
            }
        }
        return false;
    }

    private boolean dfs(int previous, int current, String number) {
        if (current + 1 == number.length()) {
            return true;
        }
        for (int i = current; i < number.length(); i++) {
            if (number.charAt(current) == '0' && i > current) {
                break;
            }
            StringBuilder sb = new StringBuilder();
            int carry = 0;
            for (int j = 0; previous + j < current || current + j <= i || carry > 0; j++) {
                int sum = carry;
                if (previous + j < current) {
                    sum += Character.getNumericValue(number.charAt(current - j - 1));
                }
                if (current + j <= i) {
                    sum += Character.getNumericValue(number.charAt(i - j));
                }
                carry = sum / 10;
                sb.insert(0, sum % 10);
            }
            //System.out.println(number.substring(previous, current) + " " + number.substring(current, i + 1) + ": " + sb.toString());
            String sum = sb.toString();
            int length = i + 1 + sum.length();
            if (length > number.length()) {
                break;
            }
            if (sum.equals(number.substring(i + 1, length)) && (length == number.length() || this.dfs(current, i + 1, number))) {
                return true;
            }
        }
        return false;
    }
}class Solution {
    public boolean isAdditiveNumber(String num) {
        if (num == null || num.length() == 0) {
            return false;
        }
        for (int i = 1; i + 1 < num.length(); i++) {
            if (this.dfs(0, i, num)) {
                return true;
            }
        }
        return false;
    }

    private boolean dfs(int previous, int current, String number) {
        if (current + 1 == number.length()) {
            return true;
        }
        for (int i = current; i < number.length(); i++) {
            if (number.charAt(current) == '0' && i > current) {
                break;
            }
            StringBuilder sb = new StringBuilder();
            int carry = 0;
            for (int j = 0; previous + j < current || current + j <= i || carry > 0; j++) {
                int sum = carry;
                if (previous + j < current) {
                    sum += Character.getNumericValue(number.charAt(current - j - 1));
                }
                if (current + j <= i) {
                    sum += Character.getNumericValue(number.charAt(i - j));
                }
                carry = sum / 10;
                sb.insert(0, sum % 10);
            }
            //System.out.println(number.substring(previous, current) + " " + number.substring(current, i + 1) + ": " + sb.toString());
            String sum = sb.toString();
            int length = i + 1 + sum.length();
            if (length > number.length()) {
                break;
            }
            if (sum.equals(number.substring(i + 1, length)) && (length == number.length() || this.dfs(current, i + 1, number))) {
                return true;
            }
        }
        return false;
    }
}